### PR TITLE
Fix flaky test with DataWriter (part 2)

### PR DESCRIPTION
**Pull request by grodowski**
Improves the `DataWriter` setup by creating threads and connecting earlier in `initialize`. Includes synchronization on `DataWriter` threads, so that calling `#start` ensures a first write to db before returning and letting the test proceed.

This attempts to improve flakiness of `test_interrupt_resume_inline_verifier_with_datawriter`, which sometimes fails as expected `BinlogVerifyStore` contents are empty on interrup, when they should contain some data from `DataWriter`.

I found out with the below log entry that nothing gets to the inline verifier before interrupt in cases where `test_interrupt_resume_inline_verifier_with_datawriter` fails (feel free to try out locally).

Hoping this will be the definitive (and correct) fix, as opposed to the speed-up from https://github.com/Shopify/ghostferry/pull/371.

---

```
diff --git a/inline_verifier.go b/inline_verifier.go
index 552c88e..b1ae817 100644
--- a/inline_verifier.go
+++ b/inline_verifier.go
@@ -25,6 +25,7 @@ import (
 // TODO: remove IterativeVerifier and remove this comment.
 type BinlogVerifyStore struct {
 	EmitLogPerRowsAdded uint64
+	logger              *logrus.Entry
 
 	mutex *sync.Mutex
 	// db => table => paginationKey => number of times it changed.
@@ -110,6 +111,7 @@ func NewBinlogVerifyStore() *BinlogVerifyStore {
 		store:               make(map[string]map[string]map[uint64]int),
 		totalRowCount:       uint64(0),
 		currentRowCount:     uint64(0),
+		logger:              logrus.WithField("tag", "BinLogVerifier"),
 	}
 }
 
@@ -127,6 +129,8 @@ func (s *BinlogVerifyStore) Add(table *TableSchema, paginationKey uint64) {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
+	s.logger.Error("Add event to store DEBUG")
+
 	_, exists := s.store[table.Schema]
 	if !exists {
 		s.store[table.Schema] = make(map[string]map[uint64]int)
```

### CODE REVIEW

File: test/helpers/data_writer_helper.rb [hunk 1]
```diff
@@ -143,5 +128,32 @@ def random_real_id(connection, table)
...
+          @logger.info("starting data writer thread #{i}")
+
+          n = 0
+          begin
+            until @stop_requested do
+              write_data(connection, &on_write)
+              @started_callback_cmd << n unless @started
+              n += 1
+              # Kind of makes the following race condition a bit better...
+              # https://github.com/Shopify/ghostferry/issues/280
```
Comment (grodowski) [inline test/helpers/data_writer_helper.rb:149]: This is still a problem 👀

File: test/helpers/data_writer_helper.rb [hunk 2]
```diff
@@ -143,5 +128,32 @@ def random_real_id(connection, table)
...
+          @logger.info("data writer thread in wait mode #{i}")
+          connection = Mysql2::Client.new(@db_config)
+          on_write = @start_cmd.pop
+          @logger.info("starting data writer thread #{i}")
+
+          n = 0
+          begin
+            until @stop_requested do
+              write_data(connection, &on_write)
+              @started_callback_cmd << n unless @started
```
Comment (grodowski) [inline test/helpers/data_writer_helper.rb:146]: Lets `#start` return after first write

File: test/helpers/data_writer_helper.rb [hunk 3]
```diff
@@ -143,5 +128,32 @@ def random_real_id(connection, table)
...
       result.first["id"]
     end
+
+    private
+
+    def start_synchronized_datawriter_threads
+      @number_of_writers.times do |i|
+        @threads << Thread.new do
+          @logger.info("data writer thread in wait mode #{i}")
+          connection = Mysql2::Client.new(@db_config)
```
[the remark this answers is not in the document]
Comment (grodowski) [reply test/helpers/data_writer_helper.rb:138]: Yup, I forgot to add it back here after an experiment with the ConnectionPool gem (but thought it's a an overkill for this test setup)

File: test/helpers/data_writer_helper.rb [hunk 4]
```diff
@@ -143,5 +128,34 @@ def random_real_id(connection, table)
...
+            until @stop_requested do
+              write_data(connection, &on_write)
+              @started_callback_cmd << n unless @started
+              n += 1
+              # Kind of makes the following race condition a bit better...
+              # https://github.com/Shopify/ghostferry/issues/280
+              sleep(0.03)
+            end
+          ensure
+            connection.close
```
[the remark this answers is not in the document]
Comment (grodowski) [reply test/helpers/data_writer_helper.rb:153]: I don't think that's right, ensures are always called after any rescues or at the end of the block.  Just moved it from [here](https://github.com/Shopify/ghostferry/pull/376/files#diff-e68bcd4db3100e0b13032fa98638b1c8c1d0fb49f174269c82962ab71139e2b8L73-L74)

**Review by mtaner — APPROVED** [review]

**Pull request merged**